packages/endo/src/archive.js

```diff
@@ -0,0 +1,158 @@
+/* global StaticModuleRecord */
+/* eslint no-shadow: 0 */
+
+import { writeZip } from "./zip.js";
+import { resolve, join } from "./node-module-specifier.js";
+import { compartmentMapForNodeModules } from "./compartmap.js";
+import { search } from "./search.js";
+import { assemble } from "./assemble.js";
+
+const { entries, fromEntries } = Object;
+
+const q = JSON.stringify;
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+const makeRecordingImportHookMaker = (
+  read,
+  baseLocation,
+  manifest,
+  errors
+) => packageLocation => {
+  packageLocation = new URL(packageLocation, baseLocation).toString();
+  return async moduleSpecifier => {
+    const moduleLocation = new URL(moduleSpecifier, packageLocation).toString();
+    const moduleBytes = await read(moduleLocation).catch(_error => undefined);
+    if (moduleBytes === undefined) {
+      errors.push(
+        `missing ${q(moduleSpecifier)} needed for package ${q(packageLocation)}`
+      );
+      return new StaticModuleRecord("// Module not found", moduleSpecifier);
+    }
+    const moduleSource = decoder.decode(moduleBytes);
+
+    const packageManifest = manifest[packageLocation] || {};
+    manifest[packageLocation] = packageManifest;
+    packageManifest[moduleSpecifier] = moduleBytes;
+
+    return new StaticModuleRecord(moduleSource, moduleLocation);
+  };
+};
+
+const renameCompartments = compartments => {
+  const renames = {};
+  let n = 0;
+  for (const [name, compartment] of entries(compartments)) {
+    const { label } = compartment;
+    renames[name] = `${label}#${n}`;
```

Comment:
Is the counter suffix a hedge against `yarn` creating multiple copies of the same `$package@$version` at different locations in the original `node_modules/` directories?

Reply:
Yes.